lib/autoproj/ops/configuration.rb

```diff
@@ -314,56 +314,56 @@ def inspect
             def sort_package_sets_by_import_order(package_sets, root_pkg_set)
-                # The sorting is done in two steps:
-                #  - first, we build a topological order of the package sets
-                #  - then, we insert the auto-imported packages, following this
-                #    topological order, in the user-provided order. Each package is
-                #    considered in turn, and added at the earliest place that fits
-                #    the dependencies
-                topological = Array.new
-                queue = (package_sets.to_a + [root_pkg_set]).uniq
-                until queue.empty?
-                    last_size = queue.size
-                    pending = queue.dup
-                    queue = Array.new
-                    until pending.empty?
-                        pkg_set = pending.shift
-                        if pkg_set.imports.any? { |imported_set| !topological.include?(imported_set) }
-                            queue.push(pkg_set)
-                        else
-                            topological << pkg_set
-                        end
-                    end
-                    if queue.size == last_size
-                        raise ArgumentError, "cannot resolve the dependencies between package sets. There seem to be a cycle amongst #{queue.map(&:name).sort.join(', ')}"
-                    end
+                name_to_pkg_set = Hash.new
+                pkg_sets = Hash.new
+                all_dependencies = []
+
+                package_sets.each do |p|
+                    name_to_pkg_set[p.name] = p
+                    pkg_sets[p.name] = p.imports.map(&:name)
+                    all_dependencies += pkg_sets[p.name]
+                end
+                root_pkg_set.imports.each do |p|
+                    pkg_sets[p.name] = p.imports.map(&:name)
+                    all_dependencies += pkg_sets[p.name]
                 end
 
-                result = root_pkg_set.imports.to_a.dup
-                to_insert = topological.dup
-                                       .find_all { |p| !result.include?(p) }
-                until to_insert.empty?
-                    pkg_set = to_insert.shift
-                    dependencies = pkg_set.imports.dup
-                    if dependencies.empty?
-                        result.unshift(pkg_set)
-                    else
-                        i = result.find_index do |p|
-                            dependencies.delete(p)
-                            dependencies.empty?
+                unknown_pkg_sets = all_dependencies - pkg_sets.keys
+                unless unknown_pkg_sets.empty?
+                    raise "Unknown package sets found: #{unknown_pkg_sets}"
+                end
+
+                sorted_pkg_sets = []
+                pending_pkg_sets = pkg_sets.keys
+
+                until pending_pkg_sets.empty?
+                    add = []
+                    progress = false
```

Comment:
I believe that progress == true is the same as !add.empty? (i.e. no need for an extra flag)